jenkinsci/jenkins
# Make widgets really pluggable

`Widget` is historically an `ExtensionPoint`, however it is customizable only for views. This makes widgets customizable as well for `Job`s, `Computer` or `ComputerSet` (current widget users).

Introduce a `HasWidgets` interface holding generic facilities to look up applicable widgets, as well as binding widgets to relative urls based on their url names. The following classes are concerned
* `View`
* `Job`
* `ComputerSet`
* `Computer`
* `CloudSet`

Introduce a generic `WidgetFactory` to allow plugins to contribute widgets to the mentioned types.

This deprecates the pattern of annotating Widget implementations with `@Extension`, in favor of typed `WidgetFactory` annotated implementation.

* Moved views `ajaxExecutors` and `ajaxBuildQueue` back to the widgets they belong to. The specific logic is now in the `WidgetFactory` implementations (as `ajax` view).

* Add a `getUrl` method to Widget. This allows to resolve custom widget views under `widget/<widgetName>` both under the Jenkins root object, jobs and views.

* `<t:queue>` and `<t:executors>` taglibs usage outside of widget context is now deprecated. Instead, implement a `WidgetFactory` passing the relevant data context.

# Plugins impact

* Subclasses of `Widget` annotated with `@Extension` still works, but is deprecated. It is advised to remove `@Extension` annotation, implement a `WidgetFactory` and annotate it with `@Extension`.
* Implement `HasWidgets` when applicable and add relevant `WidgetFactory` subclasses as needed.
* Usages of Widgets removal should use an `ExtensionFilter` removing the corresponding `WidgetFactory` implementations.



See [JENKINS-XXXXX](https://issues.jenkins.io/browse/JENKINS-XXXXX).



### Testing done

* Create a pipeline job `p` as
```
node('some-missing-label') {} 
```
and launch it
* Main Jenkins view, check content of queue (1 item), check ajax requests (successful)
* Created custom empty view, check content of queue (1 item), check ajax requests (successful). Edit the view, tick Filter Build Queue and Filter build executors. Check that queue is now empty
* Browse `/computer/`, check widgets content and ajax requests
* Browse `/computer/(built-in)/`, check widgets content and ajax requests.
* Browse `/job/p`, check build history is displayed appropriately.



### Proposed changelog entries

- Developer: plugins can now contribute widgets for `Computer`, `ComputerSet`, `View`, `Job`.



### Proposed upgrade guidelines

N/A

### Submitter checklist

  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.

### Desired reviewers

@mention



### Maintainer checklist

Before the changes are marked as `ready-for-merge`:



<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7932"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
